src/realm/object-store/sync/sync_session.cpp

```diff
@@ -883,21 +883,30 @@ static sync::Session::Config::ClientReset make_client_reset_config(RealmConfig s
 
     session_config.sync_config = std::make_shared<SyncConfig>(*session_config.sync_config); // deep copy
     session_config.scheduler = nullptr;
+    auto make_frozen_config = [](RealmConfig config) {
```

Comment:
neat! this should cover the cases we discussed: opening the realm for the first time as well as opening it after that.

Reply:
I wonder if there are any issues when reading from such a realm (I guess that's pretty much what you can do with a frozen realm). Do we have any tests?

Reply:
The unexpected thing from a user's perspective is that this Realm may not have the classes that were defined in the schema. The test does check that those don't exist. I'm not sure it would be worth making a write on a Realm that is about to be reset just to add the schema.

Reply:
Client reset may fail so in that case async open would fail too I think. The realm should end up with no schema (i.e, old schema) in this case.

Reply:
Yes, the extra classes are not present.

Reply:
But any client reset (from migration or otherwise), where classes were added to both the client and server, could cause this failure? Or would the schema versioning take care of this?

Reply:
Should we just write the schema to the realm before syncing for the async case as well? Or is it not possible?

Reply:
This is not my area of expertise, but I believe it is an optimization during async open to not write the schema. When opening a Realm for the first time in async open, the server is able to send download messages to an empty local state and have the client do OT with the schema (which is likely the same). If the client first wrote the schema, sync would send an UPLOAD and the server would have to do OT before sending any downloads. In this way, async open is supposed to be an optimization for bootstrapping.

Reply:
Yeah, the original problem it was solving was that merging the sync initialization instructions with the full sync history can be very slow. This is no longer applicable with flexible sync, but it does still apply to partition-based sync and making them work differently seems like it could greatly complicate things. Once we fully drop PBS we should probably reconsider it, though.

Reply:
Missing Tables for object types in the schema is a normal thing for read-only Realms and is something that should be totally transparent to the end user. The Empty mode on Results is there to let SDKs pretend that a missing table is merely an empty table.

Reply:
Great, thanks for the clarification @tgoyne!

Reply:
Makes sense - thank you!